scvi/dataset/dataset.py

```diff
@@ -1,431 +1,1169 @@
+            else X
+        )
+
+        self.initialize_cell_attribute(
+            "batch_indices",
+            np.asarray(batch_indices).reshape((-1, 1))
+            if batch_indices is not None
+            else np.zeros((X.shape[0], 1)),
+            categorical=True,
+        )
+        self.initialize_cell_attribute(
+            "labels",
+            np.asarray(labels).reshape((-1, 1))
+            if labels is not None
+            else np.zeros((X.shape[0], 1)),
+            categorical=True,
+        )
+
+        self.compute_library_size_batch()
 
         if gene_names is not None:
-            assert self.nb_genes == len(gene_names)
-            self.gene_names = np.array(gene_names, dtype=np.str)
+            self.initialize_gene_attribute(
+                "gene_names", np.char.upper(np.asarray(gene_names, dtype=np.str))
+            )
         if cell_types is not None:
-            assert self.n_labels == len(cell_types)
-            self.cell_types = np.array(cell_types, dtype=np.str)
+            self.initialize_mapped_attribute(
+                "labels", "cell_types", np.asarray(cell_types, dtype=np.str)
+            )
+
+        # handle additional attributes
+        if cell_attributes_dict:
+            for attribute_name, attribute_value in cell_attributes_dict.items():
+                self.initialize_cell_attribute(attribute_name, attribute_value)
+        if gene_attributes_dict:
+            for attribute_name, attribute_value in gene_attributes_dict.items():
+                self.initialize_gene_attribute(attribute_name, attribute_value)
+
+        self.remap_categorical_attributes()
+
+    def populate_from_per_batch_array(
+        self,
+        X: np.ndarray,
+        labels_per_batch: Union[np.ndarray, List[Union[List[int], np.ndarray]]] = None,
+        gene_names: Union[List[str], np.ndarray] = None,
+    ):
+        """Populates the data attributes of a GeneExpressionDataset object
+        from an array with shape (n_batches, nb_cells, nb_genes).
+
+        :param X: np.ndarray with shape (n_batches, nb_cells, nb_genes).
+        :param labels_per_batch: cell-wise labels for each batch.
+        :param gene_names: gene names, stored as ``str``.
+        """
+        if len(np.squeeze(X.shape)) != 3:
+            raise ValueError(
+                "Shape of np.squeeze(X) != 3. Use populate_from_data "
+                "if your dataset has shape (nb_cells, nb_genes)"
+            )
+        batch_indices = np.arange(X.shape[0])[:, None] * np.ones(X.shape[1])[None, :]
+        batch_indices = batch_indices.reshape(-1)
+        labels = (
+            np.concatenate(labels_per_batch).astype(np.int64)
+            if labels_per_batch is not None
+            else None
+        )
+        X = X.reshape(-1, X.shape[2])
+        self.populate_from_data(
+            X=X, batch_indices=batch_indices, labels=labels, gene_names=gene_names
+        )
+
+    def populate_from_per_batch_list(
```

Comment:
I can't find any usage of this function in the codebase (neither for its previous implementation in current master).
In addition, I don't really see a dataset which can have a shape `(n_batches, n_cells, n_genes)` as different batches usually mean different cells. 
We should remove it.

Reply:
Done